### ISSUE TITLE
Support reverse queries efficiently

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#9130]
This is an umbrella issue for improving reverse queries efficiency.

# Introduction
CQL supports the notion of reversed queries. These are queries where the clustering key ORDER BY clause specifies an order that is reversed to the one specified in the table creation clause.

Currently, Scylla supports reversed queries very inefficiently. It reads the entire data set requested (ignoring paging), reverses it in memory, and emits the first page. This can consume large amounts of memory, and if multiple pages are required, has quadratic performance in the number of pages.

The sstable format and cache/memtable data structures have everything needed for efficient reverse traversal, so it’s just a matter of wiring it.

# Motivation, goals and non-goals
## Motivation
Customers having reverse queries with a large datasets.
## Goals
- O(page) memory use
- Non-quadratic performance
- No reactor stalls
## Non-Goals
- Match the performance of forward queries
- Multi-partition reverse scans

Changes are planned in the following areas:
1. SSTable reader (Note: we probably only need to implement single-partition reversed reads.)
2. Other flat_mutation_readers
3. Cache
4. Memtable
5. Front-end

See [Supporting reversed queries efficiently](https://docs.google.com/document/d/192XATwnsv_VHc76AsEZmVkMgVh0I33OzYCFn3LQDT5o/edit) for more details.

Subordinate issues will be opened per-area/isolated task.

**Comment by denesb** [comment]
What is the relation with the old reverse query issue (https://github.com/scylladb/scylla/issues/1413). Should we close that in favor of this issue?

**Comment by amnonh** [comment]
Once this is done, Monitoring should be updated:
https://github.com/scylladb/scylla-monitoring/issues/1498

**Comment by nyh** [comment]
Also, there is issue #7586 about Alternator front-end's use of reverse queries (Alternator uses the same machinary that CQL uses, so whatever we do should benefit it as well).

**Comment by avikivity** [comment]
We should close this in favor of the original.

**Comment by bhalevy** [comment]
> We should close this in favor of the original.

Moved the additional verbiage to #1413 and closing this issue.
